developer/src/kmc-keyboard/src/compiler/layr.ts

```diff
@@ -18,28 +17,62 @@ export class LayrCompiler extends SectionCompiler {
 
   public validate() {
     let valid = true;
-    if(!this.keyboard.layers?.[0]?.layer?.length) {
+    if (!this.keyboard.layers?.[0]?.layer?.length) {
       valid = false;
       this.callbacks.reportMessage(CompilerMessages.Error_MustBeAtLeastOneLayerElement());
     }
-    // TODO-LDML
+    let hardwareLayers = 0;
+    this.keyboard.layers.forEach(({ hardware, form }) => {
+      // TODO-LDML: in the future >1 hardware layer may be allowed, check for duplicates
+      if (form === 'touch') {
+        if (hardware) {
+          valid = false;
+          this.callbacks.reportMessage(CompilerMessages.Error_InvalidFile({
+            errorText: `Not allowed: hardware="${hardware}" with layers form="touch"`
+          }));
+        }
+      } else if (form === 'hardware') {
+        hardwareLayers++;
+        if (!hardware) {
+          valid = false;
+          this.callbacks.reportMessage(CompilerMessages.Error_InvalidFile({
+            errorText: `on layers form="hardware", missing required hardware= attribute.`
+          }));
+        } else if (!constants.layr_list_hardware_map.get(hardware)) {
+          valid = false;
+          this.callbacks.reportMessage(CompilerMessages.Error_InvalidFile({
+            errorText: `Unknown hardware layout id: hardware="${hardware}"`
+          }));
+        } else if (hardwareLayers > 1) { // TODO-LDML: revisit if spec changes
+          valid = false;
+          this.callbacks.reportMessage(CompilerMessages.Error_MustHaveAtMostOneLayersElementPerForm({ form }));
+        }
```

Comment:
We really want to have separate message identifiers for each of these rather than using `Error_InvalidFile` repeatedly, which also moves the error text to the `CompilerMessages` unit, for future localization. (I know that we are refactoring form vs hardware so some of this may go away...)

Reply:
I think for that last reason I'd like to make this a TODO-LDML since the identifier will also change at that time. 

Anyway, revamped this.  I don't mind busywork. But if I can use the busywork to improve things, I like to do that.